### ISSUE TITLE
[ENH] investigate use of abc for base classes, write STEP

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#1923]
@TonyBagnall recently suggested to use of abc for classifiers in *#1873, #1708), and this was also previously discussed a number of times: #1424, #1708, #510, #420.

I do think we need to work out a STEP for this, it would be a major change to the interface framework.

There are some open questions that I have:
* which classes would inherit from `ABCMeta`?
* how do we deal with optional abstract methods?
* how does this integrate with the tag system in `BaseObject`? This could be updated using `ChainMap`.
* are there any problematic interactions of the public/private nesting pattern `fit`/`_fit`?
* we would still like to have capability tags that tell us whether an optional method is implemented or not, e.g., `predict_proba` or `predict_interval`. How would this interact with `abc`?
* how would downwards compatibility and deprecation work here?
* what about polymorphic estimators, e.g., a pipeline that can act like a transformer or a forecaster? Would `abc` not complain about missing transformer methods if it happens to behave like a forecaster and vice versa?

There's probably more once we have a proper design discussion, this is only off the top of my head.

**Comment by chrisholder** [comment]
Why is this an issue. This is the standard best practice way of doing abstract methods in python why wouldn't we do it.

**Comment by chrisholder** [comment]
To answer your questions more directly though I think you're overcomplicating  a lot of it. Literally it will feel the same as the current raise NotImplementedError except it's standardised and more util methods (and ides can better support auto implement functionality):

- which classes would inherit from ABCMeta?
BaseClusterer, BaseClassifier, BaseForecaster etc. Essentially any estimator/transformer that current uses raise NotImplementedError(). This is because ABC just does what that does.

- how do we deal with optional abstract methods?
Optional abstract methods aren't really abstract method. What I think you are wanting is just method overriding. So add a default behaviour and override it in the parent class when you want to change the default behaviour.

- how does this integrate with the tag system in BaseObject? This could be updated using ChainMap.
This wouldn't touch any of this?

- are there any problematic interactions of the public/private nesting pattern fit/_fit?
No.

- we would still like to have capability tags that tell us whether an optional method is implemented or not, e.g., predict_proba or predict_interval. How would this interact with abc?
It doesn't touch it. It's an abstract method it doesn't touch functionality.

- how would downwards compatibility and deprecation work here?
Abc has been supported since 3.6. Again you don't deprecate anything you're just moving away from raise NotImplementedError().

- what about polymorphic estimators, e.g., a pipeline that can act like a transformer or a forecaster? Would abc not complain - - about missing transformer methods if it happens to behave like a forecaster and vice versa?
These aren't abstract methods. But I know they exist and where this happens just raise NotImplementedError. We shouldn't be not using the best practise stuff where it's appropriate just because a niche scenario it doesn't work.

**Comment by TonyBagnall** [comment]
1. which classes would inherit from ABCMeta?
ones with abstract methods currently raising errors if not implemented

2. how do we deal with optional abstract methods?
this has no meaning. The definition of an abstract method is one a subclass *must* implement. Either there is sensible default behaviour, or its abstract. If you want to still raise errors for optional methods, or just pass if it makes sense, thats fine. ABC doesnt stop you doing this. 

3. how does this integrate with the tag system in BaseObject? This could be updated using ChainMap.
not sure why making methods abstract would influence tags at all

4. are there any problematic interactions of the public/private nesting pattern fit/_fit?
how can there be? In fact it enforces correct behaviour at compile time, thats the whole point

5.we would still like to have capability tags that tell us whether an optional method is implemented or not, e.g., predict_proba or predict_interval. How would this interact with abc?
if you want a method optional, it should not be abstract. You just carry on as before. 
>how would downwards compatibility and deprecation work here?
see chris's answer

>what about polymorphic estimators, e.g., a pipeline that can act like a transformer or a forecaster? Would abc not complain about missing transformer methods if it happens to behave like a forecaster and vice versa?
dont make them abstract unless the *have* to be implemented by subclasses. I honestly dont understand you points, I think you have misunderstood the meaning of abstract methods, or at least the meaning we teach our first years. If you want to use them in some weird non standard way for forecasting, go for it, maybe there is even a python histroy of it, I dont know. However, I dont know why you are persistently blocking such a simple, uncontroversial change for classifiers when ABC is already used in benchmarking (which you managed), and why you did not object to its use in clustering and distances.

**Comment by fkiraly** [comment]
> which classes would inherit from ABCMeta?
BaseClusterer, BaseClassifier, BaseForecaster etc. Essentially any estimator/transformer that current uses raise NotImplementedError(). This is because ABC just does what that does.

I worry a bit about the overhead this causes.
Should not `BaseObject` inherit from it, and nothing else?

> Optional abstract methods aren't really abstract method. What I think you are wanting is just method overriding. So add a default behaviour and override it in the parent class when you want to change the default behaviour.

No, this won't work - there are methods that have no sensible default behaviour and should raise an error. Examples: `predict_interval` for forecasters.

>> how would downwards compatibility and deprecation work here?
> Abc has been supported since 3.6.

Your answer is besides the point: I mean, downwards compatibility with `sktime` itself before we make the change.
Are we certain this won't break custom estimators, for instance? If so, what is the argument to be certain?

>>what about polymorphic estimators, e.g., a pipeline that can act like a transformer or a forecaster? 
>We shouldn't be not using the best practise stuff where it's appropriate just because a niche scenario it doesn't work.

Very shaky argument imo. Polymorphic pipelines are not nice, the pipeline is in fact a central object in `sklearn`-like libraries. I am against abc unless we can ensure that polymorphic pipelines such as the `sklearn` one work properly.

**Comment by fkiraly** [comment]
@TonyBagnall 

> I think you have misunderstood the meaning of abstract methods, or at least the meaning we teach our first years. 

I know what abstract methods are and how they work in abc.
A descendant will exception if it doesn't override all the abstracts declared by abc.

>> which classes would inherit from ABCMeta?
>ones with abstract methods currently raising errors if not implemented

Why not `BaseObject`? I don't like that we inherit from multiple classes again.

>> how do we deal with optional abstract methods?
> this has no meaning. [...] Either there is sensible default behaviour, or its abstract. 

Disagreed, I think you don't get the conceptual point here: some base types have methods that have no sensible default *and* descendants do not need to implement them, so cannot be abstract. Example is `predict_quantiles` for forecasters.

**Comment by fkiraly** [comment]
overall, I'm happy with answers to 3, 4, 5. I'm not happy and concerned about the others.

**Comment by TonyBagnall** [comment]
Ultimately I don't really mind, if you object we won't do it, but it seems a really odd block to me, clearly missing some nuance in your design. So do you want to strip ABC out of the code base?

**Comment by fkiraly** [comment]
>Ultimately I don't really mind, if you object we won't do it, but it seems a really odd block to me, clearly missing some nuance in your design. So do you want to strip ABC out of the code base? If so, could you start with benchmarking and then we will follow

I'm actually in favour since it's elegant. I just don't want to miss an important problem this may cause, such as complicating the extender contract.

**Comment by TonyBagnall** [comment]
I'm really confused now, you blocked my pr making classifier abstract but want it? I'll reopen the PR

**Comment by fkiraly** [comment]
> I'm really confused now, you blocked my pr making classifier abstract but want it? I'll reopen the PR

Sorry for not being clear:
* I like the general idea and am quite supportive of it
* I think we need to clarify some open design questions before we can merge the concrete PR

This is not a contradiction! All I want is us avoiding waking up to a bad surprise because we did not think the design through.
I still consider some of my questions above open or not sufficiently answered.

E.g., how does the extender contract concretely change?

**Comment by fkiraly** [comment]
I think we have implicitly decided not to use abc, becaus:

* it clashes with inheritance patterns in which some methods that are abstract should be overridden only in conditional cases, depending on tags. For instance, `BaseForecaster._predict_interval` etc should only be implemented if the `capability:pred_int` tag is set.
* it clashes with the pattern of having a default private logic that uses other methods if not overridden, e.g., getting `)predict_quantiles` from `_predict_interval` or vice versa in `BaseForecaster`

Also see another example of this problem here where we had to remove `abc`: https://github.com/sktime/sktime/pull/4527
